File: SDK/Solutions/VB/Exercise3/MyAnimal.vb

<Assembly: OrganismClass("Exercise3.MyAnimal")> 
<Assembly: AuthorInformation("Your Name", "someone@microsoft.com")> 

< _
    Carnivore(False), _
    MatureSize(36), _
    AnimalSkin(AnimalSkinFamily.Beetle), _
    MarkingColor(KnownColor.Red), _
    MaximumEnergyPoints(20), _
    EatingSpeedPointsAttribute(0), _
    AttackDamagePointsAttribute(12), _
    DefendDamagePointsAttribute(12), _
    MaximumSpeedPointsAttribute(16), _
    CamouflagePointsAttribute(10), _
    EyesightPointsAttribute(20) _
> _
Public Class MyAnimal : Inherits Animal

    Dim targetPlant As PlantState = Nothing    ' The current plant we're going after
    Protected Overloads Overrides Sub Initialize()

        ' TODO: Add Initialization logic here
        AddHandler Idle, AddressOf MyAnimal_Idle
        AddHandler Load, AddressOf MyAnimal_Load
        AddHandler Attacked, AddressOf MyAnimal_Attacked
        AddHandler MoveCompleted, AddressOf MyAnimal_MoveCompleted

    End Sub

    ' First event fired on an organism each turn 
    Sub MyAnimal_Load(ByVal sender As Object, ByVal e As LoadEventArgs)
        Try
            If Not (targetPlant Is Nothing) Then
                ' See if our target plant still exists (it may have died) 
                ' LookFor returns null if it isn't found 
                targetPlant = CType(LookFor(targetPlant), PlantState)
                If (targetPlant Is Nothing) Then

                    ' WriteTrace is the best way to debug your creatures. 
                    WriteTrace("Target plant disappeared.")
                End If
            End If
        Catch exc As Exception
            WriteTrace(exc.ToString())
        End Try
    End Sub

    ' Fired if we're being attacked 
    Sub MyAnimal_Attacked(ByVal sender As Object, ByVal e As AttackedEventArgs)
        If (e.Attacker.IsAlive) Then
            Dim TheAttacker As AnimalState = e.Attacker
            BeginDefending(TheAttacker) 'defend against the attacker 

            WriteTrace("Run away to some random point")
            Dim x As Integer = OrganismRandom.Next(0, WorldWidth - 1)
            Dim y As Integer = OrganismRandom.Next(0, WorldHeight - 1)
            BeginMoving(New MovementVector(New Point(x, y), 10))
        End If
    End Sub

    ' Fired when we've finished moving.
    Sub MyAnimal_MoveCompleted(ByVal sender As Object, ByVal e As MoveCompletedEventArgs)
        ' Reset the antenna value
        Antennas.AntennaValue = 0

        ' If we've stopped because something is blocking us...
        If (e.Reason = ReasonForStop.Blocked) Then
            WriteTrace("Something's blocking my way.")
            If (TypeOf e.BlockingOrganism Is AnimalState) Then
                Dim blockingAnimal As AnimalState = CType(e.BlockingOrganism, AnimalState)
                If (blockingAnimal.AnimalSpecies.IsSameSpecies(Me.Species)) Then
                    ' Signal to our friend to move out of our way.
                    WriteTrace("One of my friends is blocking my way.  I'll ask him to move.")
                    Antennas.AntennaValue = 13
                End If
            End If
        End If
    End Sub

    ' Fired after all other events are fired during a turn 
    Sub MyAnimal_Idle(ByVal sender As Object, ByVal e As IdleEventArgs)
        Try
            ' Reproduce as often as possible 
            If (CanReproduce) Then
                BeginReproduction(Nothing)
            End If

            ' If we can eat and we have a target plant, eat 
            If (CanEat) Then
                WriteTrace("Hungry.")
                If Not (IsEating) Then
                    WriteTrace("Not eating: Have target plant?")
                    If Not (targetPlant Is Nothing) Then
                        WriteTrace("Yes, Have target plant already.")
                        If (WithinEatingRange(targetPlant)) Then
                            WriteTrace("Within Range, Start eating.")
                            BeginEating(targetPlant)
                            If (IsMoving) Then
                                WriteTrace("Stop while eating.")
                                StopMoving()
                            End If
                        Else
                            If Not (IsMoving) Then
                                WriteTrace("Move to Target Plant")
                                BeginMoving(New MovementVector(targetPlant.Position, 2))
                            End If
                        End If
                    Else
                        WriteTrace("Don't have target plant.")
                        If Not (ScanForTargetPlant()) Then
                            If Not (IsMoving) Then
                                WriteTrace("No plant found, so pick a random point and move there")
                                Dim RandomX As Integer = OrganismRandom.Next(0, WorldWidth - 1)
                                Dim RandomY As Integer = OrganismRandom.Next(0, WorldHeight - 1)
                                BeginMoving(New MovementVector(New Point(RandomX, RandomY), 2))
                            Else
                                WriteTrace("Moving and Looking...")
                            End If
                        End If
                    End If
                Else
                    WriteTrace("Eating.")
                    If (IsMoving) Then
                        WriteTrace("Stop moving while eating.")
                        StopMoving()
                    End If
                End If
            Else
                WriteTrace("Full: do nothing.")
                If (IsMoving) Then
                    StopMoving()
                End If
            End If
            ShouldIMoveForMyFriend()

        Catch exc As Exception
            WriteTrace(exc.ToString())
        End Try
    End Sub

    ' Looks for target plants, and starts moving towards the first one it finds 
    Function ScanForTargetPlant() As Boolean
        Try
            Dim foundCreatures As System.Collections.ArrayList = Scan()
            If (foundCreatures.Count > 0) Then
                Dim orgState As OrganismState
                ' Always move after closest plant or defend closest creature if there is one 
                For Each orgState In foundCreatures
                    If (TypeOf orgState Is PlantState) Then
                        targetPlant = CType(orgState, PlantState)
                        BeginMoving(New MovementVector(orgState.Position, 2))
                        Return True
                    End If
                Next
            End If
        Catch exc As Exception
            WriteTrace(exc.ToString())
        End Try
        Return False
    End Function

    ' Routine to move out of the way when blocking our friends.
    Sub ShouldIMoveForMyFriend()
        Try
            Dim foundAnimals As System.Collections.ArrayList = Scan()
            If (foundAnimals.Count > 0) Then

                Dim orgState As OrganismState
                For Each orgState In foundAnimals
                    If (TypeOf orgState Is AnimalState) Then
                        Dim visibleAnimal As AnimalState = CType(orgState, AnimalState)
                        ' Only move if the animal is one of our friends (IsSameSpecies).
                        If (visibleAnimal.Species.IsSameSpecies(Me.Species)) Then
                            ' If the animal's antenna value is 13, it means they're blocked (see the MoveCompletedEvent method).
                            If (visibleAnimal.Antennas.AntennaValue = 13) Then
                                ' We're blocking our friend, so we should move.
                                WriteTrace("I'm blocking one of my friends.  I should move.")

                                Dim newX As Integer = Me.Position.X - (visibleAnimal.Position.X - Me.Position.X)
                                Dim newY As Integer = Me.Position.Y - (visibleAnimal.Position.Y - Me.Position.Y)

                                BeginMoving(New MovementVector(New Point(newX, newY), 2))
                                Return
                            End If
                        End If
                    End If
                Next
            End If

        Catch exc As Exception
            WriteTrace(exc.ToString())
        End Try
    End Sub

    Public Overloads Overrides Sub SerializeAnimal(ByVal m As MemoryStream)

        ' TODO: Add Serialization logic here

    End Sub

    Public Overloads Overrides Sub DeserializeAnimal(ByVal m As MemoryStream)

        ' TODO: Add Deserialization logic here

    End Sub

End Class

File: SDK/Solutions/CS/Exercise3/MyAnimal.cs
using System;
using System.Collections;
using System.Drawing;
using System.IO;
using OrganismBase;

[assembly : OrganismClass("Exercise3.MyAnimal")]
[assembly : AuthorInformation("Your Name", "someone@microsoft.com")]

namespace Exercise3
{
    [Carnivore(false)]
    [MatureSize(28)]
    [AnimalSkin(AnimalSkinFamily.Inchworm)]
    [MarkingColor(KnownColor.Yellow)]
    [MaximumEnergyPoints(20)]
    [EatingSpeedPointsAttribute(0)]
    [AttackDamagePointsAttribute(12)]
    [DefendDamagePointsAttribute(12)]
    [MaximumSpeedPointsAttribute(16)]
    [CamouflagePointsAttribute(10)]
    [EyesightPointsAttribute(20)]
    public class MyAnimal : Animal
    {
        private PlantState targetPlant = null; // The current plant we're going after

        protected override void Initialize()
        {
            // TODO: Add initialization logic here
            Idle += MyAnimal_Idle;
            Load += MyAnimal_Load;
            Attacked += MyAnimal_Attacked;
            MoveCompleted += MyAnimal_MoveCompleted;
        }

        // First event fired on an organism each turn 
        private void MyAnimal_Load(object sender, LoadEventArgs e)
        {
            try
            {
                if (targetPlant != null)
                {
                    // See if our target plant still exists (it may have died) 
                    // LookFor returns null if it isn't found 
                    targetPlant = (PlantState) LookFor(targetPlant);
                    if (targetPlant == null)

                        // WriteTrace is the best way to debug your creatures. 
                        WriteTrace("Target plant disappeared.");
                }
            }
            catch (Exception exc)
            {
                WriteTrace(exc.ToString());
            }
        }

        // Fired if we're being attacked 
        private void MyAnimal_Attacked(object sender, AttackedEventArgs e)
        {
            if (e.Attacker.IsAlive)
            {
                AnimalState TheAttacker = e.Attacker;

                BeginDefending(TheAttacker); //defend against the attacker 
                WriteTrace("Run away to some random point");

                int X = OrganismRandom.Next(0, WorldWidth - 1);
                int Y = OrganismRandom.Next(0, WorldHeight - 1);

                BeginMoving(new MovementVector(new Point(X, Y), 10));
            }
        }

        // Fired when we've finished moving.
        private void MyAnimal_MoveCompleted(object sender, MoveCompletedEventArgs e)
        {
            // Reset the antenna value
            Antennas.AntennaValue = 0;

            // If we've stopped because something is blocking us...
            if (e.Reason == ReasonForStop.Blocked)
            {
                WriteTrace("Something's blocking my way.");
                if (e.BlockingOrganism is AnimalState)
                {
                    AnimalState blockingAnimal = (AnimalState) e.BlockingOrganism;

                    if (blockingAnimal.AnimalSpecies.IsSameSpecies(Species))
                    {
                        // Signal to our friend to move out of our way.
                        WriteTrace("One of my friends is blocking my way.  I'll ask him to move.");
                        Antennas.AntennaValue = 13;
                    }
                }
            }
        }

        // Fired after all other events are fired during a turn 
        private void MyAnimal_Idle(object sender, IdleEventArgs e)
        {
            try
            {
                // Reproduce as often as possible 
                if (CanReproduce)
                    BeginReproduction(null);

                // If we can eat and we have a target plant, eat 
                if (CanEat)
                {
                    WriteTrace("Hungry.");
                    if (!IsEating)
                    {
                        WriteTrace("Not eating: Have target plant?");
                        if (targetPlant != null)
                        {
                            WriteTrace("Yes, Have target plant already.");
                            if (WithinEatingRange(targetPlant))
                            {
                                WriteTrace("Within Range, Start eating.");
                                BeginEating(targetPlant);
                                if (IsMoving)
                                {
                                    WriteTrace("Stop while eating.");
                                    StopMoving();
                                }
                            }
                            else
                            {
                                if (!IsMoving)
                                {
                                    WriteTrace("Move to Target Plant");
                                    BeginMoving(new MovementVector(targetPlant.Position, 2));
                                }
                            }
                        }
                        else
                        {
                            WriteTrace("Don't have target plant.");
                            if (!ScanForTargetPlant())
                                if (!IsMoving)
                                {
                                    WriteTrace("No plant found, so pick a random point and move there");

                                    int RandomX = OrganismRandom.Next(0, WorldWidth - 1);
                                    int RandomY = OrganismRandom.Next(0, WorldHeight - 1);

                                    BeginMoving(new MovementVector(new Point(RandomX, RandomY), 2));
                                }
                                else
                                {
                                    WriteTrace("Moving and Looking...");
                                }
                        }
                    }
                    else
                    {
                        WriteTrace("Eating.");
                        if (IsMoving)
                        {
                            WriteTrace("Stop moving while eating.");
                            StopMoving();
                        }
                    }
                }
                else
                {
                    WriteTrace("Full: do nothing.");
                    if (IsMoving)
                        StopMoving();
                }

                ShouldIMoveForMyFriend();
            }
            catch (Exception exc)
            {
                WriteTrace(exc.ToString());
            }
        }

        // Looks for target plants, and starts moving towards the first one it finds 
        private bool ScanForTargetPlant()
        {
            try
            {
                ArrayList foundCreatures = Scan();

                if (foundCreatures.Count > 0)
                {
                    // Always move after closest plant or defend closest creature if there is one 
                    foreach (OrganismState organismState in foundCreatures)
                    {
                        if (organismState is PlantState)
                        {
                            targetPlant = (PlantState) organismState;
                            BeginMoving(new MovementVector(organismState.Position, 2));
                            return true;
                        }
                    }
                }
            }
            catch (Exception exc)
            {
                WriteTrace(exc.ToString());
            }
            return false;
        }

        // Routine to move out of the way when blocking our friends.
        private void ShouldIMoveForMyFriend()
        {
            try
            {
                ArrayList foundAnimals = Scan();

                if (foundAnimals.Count > 0)
                {
                    foreach (OrganismState organismState in foundAnimals)
                    {
                        if (organismState is AnimalState)
                        {
                            AnimalState visibleAnimal = (AnimalState) organismState;

                            // Only move if the animal is one of our friends (IsSameSpecies).
                            if (visibleAnimal.Species.IsSameSpecies(Species))
                            {
                                // If the animal's antenna value is 13, it means they're blocked (see the MoveCompletedEvent method).
                                if (visibleAnimal.Antennas.AntennaValue == 13)
                                {
                                    // We're blocking our friend, so we should move.
                                    WriteTrace("I'm blocking one of my friends.  I should move.");

                                    int newX = Position.X - (visibleAnimal.Position.X - Position.X);
                                    int newY = Position.Y - (visibleAnimal.Position.Y - Position.Y);

                                    BeginMoving(new MovementVector(new Point(newX, newY), 2));
                                    return;
                                }
                            }
                        }
                    }
                }
            }
            catch (Exception exc)
            {
                WriteTrace(exc.ToString());
            }
        }

        public override void SerializeAnimal(MemoryStream stream)
        {
            // TODO: Add serialization logic here
        }

        public override void DeserializeAnimal(MemoryStream stream)
        {
            // TODO: Add deserialization logic here
        }
    }
}